addons/binding/org.openhab.binding.freebox/src/main/java/org/openhab/binding/freebox/internal/FreeboxAirPlayAudioSink.java

```diff
@@ -0,0 +1,119 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This makes a AirPlay device to serve as an {@link AudioSink}-
+ *
+ * @author Laurent Garnier - Initial contribution for AudioSink and notifications
+ */
+public class FreeboxAirPlayAudioSink implements AudioSink {
+
+    private final Logger logger = LoggerFactory.getLogger(FreeboxAirPlayAudioSink.class);
+
+    private static HashSet<AudioFormat> supportedFormats = new HashSet<>();
+
+    static {
+        supportedFormats.add(AudioFormat.WAV);
+        supportedFormats.add(AudioFormat.MP3);
+    }
+
+    private AudioHTTPServer audioHTTPServer;
+    private FreeboxThingHandler handler;
+    private String callbackUrl;
+
+    public FreeboxAirPlayAudioSink(FreeboxThingHandler handler, AudioHTTPServer audioHTTPServer, String callbackUrl) {
+        this.handler = handler;
+        this.audioHTTPServer = audioHTTPServer;
+        this.callbackUrl = callbackUrl;
+    }
+
+    @Override
+    public String getId() {
+        return handler.getThing().getUID().toString();
+    }
+
+    @Override
+    public String getLabel(Locale locale) {
+        return handler.getThing().getLabel();
+    }
+
+    @Override
+    public void process(AudioStream audioStream) throws UnsupportedAudioFormatException {
+        if (!ThingHandlerHelper.isHandlerInitialized(handler)
+                || (handler.getThing().getStatus() != ThingStatus.ONLINE)) {
+            return;
+        }
+
+        String url = null;
+        if (audioStream instanceof URLAudioStream) {
+            // it is an external URL, we can access it directly
+            URLAudioStream urlAudioStream = (URLAudioStream) audioStream;
+            url = urlAudioStream.getURL();
+            try {
+                audioStream.close();
+            } catch (IOException e) {
+            }
+        } else {
+            if (callbackUrl != null) {
+                // we serve it on our own HTTP server
+                String relativeUrl;
+                if (audioStream instanceof FixedLengthAudioStream) {
+                    relativeUrl = audioHTTPServer.serve((FixedLengthAudioStream) audioStream, 20);
+                } else {
+                    relativeUrl = audioHTTPServer.serve(audioStream);
+                }
+                url = callbackUrl + relativeUrl;
+            } else {
+                logger.warn("We do not have any callback url, so AirPlay device cannot play the audio stream!");
+            }
+        }
+        try {
+            logger.debug("AirPlay audio sink: process url {}", url);
+            handler.playMedia(url);
+        } catch (Exception e) {
+            logger.warn("Audio stream playback failed", e);
+        }
+    }
+
+    @Override
+    public Set<AudioFormat> getSupportedFormats() {
+        return supportedFormats;
+    }
+
+    @Override
+    public PercentType getVolume() throws IOException {
+        throw new IOException("Volume can not be determined");
```

Comment:
Yes, according to the documentation of the [IOException](https://docs.oracle.com/javase/7/docs/api/java/io/IOException.html) its about *'exceptions produced by failed or interrupted I/O operations'*

The [UnsupportedOperationException](https://docs.oracle.com/javase/7/docs/api/java/lang/UnsupportedOperationException.html) is about 'Thrown to indicate that the requested operation is not supported.'

You can always throw any runtime exception and you should throw the most specific one.